glean_parser/templates/go_server.jinja2

```diff
@@ -0,0 +1,263 @@
+	Events     []gleanEvent `json:"events"`
+}
+
+type gleanEvent struct {
+	Category  string     `json:"category"`
+	Name      string     `json:"name"`
+	Timestamp int64      `json:"timestamp"`
+	Extra     [][]string `json:"extra"`
+}
+
+func (g GleanEventsLogger) createClientInfo() clientInfo {
+	// Fields with default values are required in the Glean schema, but not used in server context
+	return clientInfo{
+		TelemetrySDKBuild: "{glean_parser_version}",
+		FirstRunDate:      "Unknown",
+		OS:                "Unknown",
+		OSVersion:         "Unknown",
+		Architecture:      "Unknown",
+		AppBuild:          "Unknown",
+		AppDisplayVersion: g.AppDisplayVersion,
+		AppChannel:        g.AppChannel,
+	}
+}
+
+func createPingInfo() pingInfo {
+	// TODO: confirm expected date format in logs,
+	// js uses Date().toISOString(), ex. "2023-12-13T20:21:55.390Z"
+	// ruby uses Time.now.utc, ex. "2023-12-13 20:21:02 UTC"
+	var now = time.Now().UTC().Format("2006-01-02 15:04:05 MST")
+	return pingInfo{
+		Seq:       0,
+		StartTime: now,
+		EndTime:   now,
+	}
+}
+
+func (g GleanEventsLogger) createPing(documentType string, config RequestInfo, payload string) ping {
+	var documentId = uuid.New()
+	return ping{
+		DocumentNamespace: g.AppID,
+		DocumentType:      documentType,
+		DocumentVersion:   "1",
+		DocumentID:        documentId.String(),
+		UserAgent:         config.UserAgent,
+		IpAddress:         config.IpAddress,
+		Payload:           string(payload),
+	}
+}
+
+// method called by each event or custom ping method
+func (g GleanEventsLogger) record(
+	documentType string,
+	requestInfo RequestInfo,
+	metrics metrics,
+	events []gleanEvent,
+) {
+	var telemetryPayload = pingPayload{
+		ClientInfo: g.createClientInfo(),
+		PingInfo:   createPingInfo(),
+		Metrics:    metrics,
+		Events:     events,
+	}
+
+	var payloadJson, payloadErr = json.Marshal(telemetryPayload)
+	if payloadErr != nil {
+		panic("Unable to marshal payload to json")
+	}
+
+	// TODO: confirm expected shape with data team
+	// ruby implementation wraps ping in another js object before JSON.dumping to a string
+	// ex. { "Timestamp": string, "Logger": name, "Type": gleanEventMozLogType, "Severity": string, "Pid": string, "Fields": pingJson }
+	// js appears not to
+	var ping = g.createPing(documentType, requestInfo, string(payloadJson))
+	var pingJson, err = json.Marshal(ping)
+	if err != nil {
+		panic("Unable to marshal ping to json")
+	}
+	g.Logger.log(gleanEventMozlogType, string(pingJson))
+}
+
+{% for ping, metrics_by_type in pings.items() %}
+{% if ping != "events" %}
+type {{ ping|ping_type_name }} struct {
+  {% for metric_type, metrics in metrics_by_type.items() %}
+  {% if metric_type != 'event' %}
+  {% for metric in metrics %}
+  {{ metric|metric_argument_name }} {{ metric.type|go_metric_type }} // {{ metric.description|clean_string }}
+  {% endfor %}
+  {% endif %}
+  {% endfor %}
+	{# TODO: if custom pings need to support including any event metric #}
```

Comment:
I noticed in the JS code that the custom pings logging method includes a param for passing in events, but if we're sending each event individually on their own in the server implementation I wasn't sure if it is really needed to send a custom ping and an event(s) at the same time.

Reply:
We chatted on slack and it is preferred that server implementations only support the events ping. The JS implementation supported either and then the ruby implementation only supported events, but that was an intentional decision which we should keep consistent with for go.

It simplifies the implementations and eases querying of the data, since each custom ping could include any of the event metrics defined, but the server implementations by default are sending the event metrics in the events ping.

As we define what data we want to collect, we can discuss if this causes any hurdles.